HeterogeneousCore/AlpakaTest/plugins/alpaka/XyzIdAlpakaProducer.cc

```diff
@@ -0,0 +1,76 @@
+#include <optional>
+#include <string>
+
+#include <alpaka/alpaka.hpp>
+
+#include "DataFormats/XyzId/interface/alpaka/XyzIdDeviceCollection.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/host.h"
+#include "HeterogeneousCore/AlpakaServices/interface/alpaka/AlpakaService.h"
+
+#include "XyzIdAlgo.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  class XyzIdAlpakaProducer : public edm::stream::EDProducer<> {
+  public:
+    XyzIdAlpakaProducer(edm::ParameterSet const& config)
+        : deviceToken_{produces<XyzIdDeviceCollection>()}, size_{config.getParameter<int32_t>("size")} {}
+
+    void beginStream(edm::StreamID sid) override {
+      // choose a device based on the EDM stream number
+      edm::Service<ALPAKA_TYPE_ALIAS(AlpakaService)> service;
+      if (not service->enabled()) {
+        throw cms::Exception("Configuration") << ALPAKA_TYPE_ALIAS_NAME(AlpakaService) << " is disabled.";
```

Comment:
Just a comment on this pattern (likely) beyond this PR. If an application has explicitly EDProducer instances for two (or more) backends, e.g. CPU serial and CUDA, and the job has an OutputModule that explicitly consumes both of them (directly or via intermediate producers, like transfer-to-host), like the HLT menu currently does, the `beginStream()` (or all non-event methods) is called for all such backends, even if a backend would be disabled in the configuration.

Reply:
You mean that `SomeCUDAEDProducer::beginStream()` is called even if `process.options.accelerators = [ "cpu" ]` ?

Reply:
Yes (in a specific situation). This is a result of the interplay of several features:
1. `process.options.accelerators` impacts only on the behavior of `SwitchProducer`s
2. framework deletes unscheduled EDModules whose data products are not consumed by any other EDModule that would otherwise be run
3. framework calls all non-event transition functions (such as beginStream, beginRun, beginLuminosityBlock) for all EDModules that remain in the system after 2)

Therefore, if an EDModule's data products are consumed unconditionally (i.e. anything not part of a dependence chain originating from a `SwitchProducer`, e.g. by `OutputModule`), the EDModule's non-event transition functions will be always run (the event transition function might still not get run, e.g. if the EDModule is in a Path behind an event-rejecting EDFilter).

In a sense in such a case the job configuration is inconsistent (for the present system): it contains a module using an accelerator that explicitly ends up being run (either being in a Path or output being consumed), but is then configured to not use the accelerator.

(from another point of view I would consider the behavior a bit annoying though, it bit me when preparing #38145)

Reply:
> Therefore, if an EDModule's data products are consumed unconditionally (i.e. anything not part of a dependence chain originating from a `SwitchProducer`, e.g. by `OutputModule`), the EDModule's non-event transition functions will be always run (the event transition function might still not get run, e.g. if the EDModule is in a Path behind an event-rejecting EDFilter).

OK, now I'm confused... I'll need to review this at a more reasonable hour.

Reply:
To add to the confusion:
  - is it the case also if the `OutputModule` is in a `FinalPath` ?
  - is it the case also if the `EDProducer` is in a `ConditionalTask` ?

Reply:
> * is it the case also if the `OutputModule` is in a `FinalPath` ?

Yes. The only impact of an `OutputModule` being in a `FinalPath` is that the `OutputModule`'s `keep` statements do not cause the modules event transition functions (`produce()`, `filter()`) to run. 

> * is it the case also if the `EDProducer` is in a `ConditionalTask` ?

Yes. The impact of `ConditionalTask` (currently) is that a module there either gets inserted in the `Path` containing the `ConditionalTask` (if it is consumed by a module that already is in the `Path`) or the module gets treated as unscheduled module (if it is not consumed by anything in the `Path`). After that the module gets treated as other scheduled or unscheduled modules. In the unscheduled case, only if no scheduled module (directly or via a chain of dependencies of (un)scheduled modules) consumes a product of the unscheduled module, the unscheduled module is removed from the system before beginJob.